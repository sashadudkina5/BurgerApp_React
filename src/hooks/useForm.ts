import React from "react";

/**
 * Custom React hook that manages form state. 
 * It initializes form fields with default values and provides a method to update these values based on user input.
 * 
 * @param {Object} inputValues - Initial values for form fields with keys representing the field names and their corresponding values. 
 * Default values are for email, password, name, token, and newPassword fields.
 * 
 * @returns {Object} An object containing:
 * - values: An object holding the current state of form values.
 * - handleChange: A function to be called on input change events, updating the form values state with the input's current value.
 * - setValues: A function to manually set the form values.
 * 
 * @example
 * const { values, handleChange, setValues } = useForm({ email: '', password: '' });
 * 
 * <input
 *   name="email"
 *   value={values.email}
 *   onChange={handleChange}
 * />
 * 
 * <input
 *   name="password"
 *   value={values.password}
 *   onChange={handleChange}
 * />
 */
export function useForm(inputValues = { email: "", password: "", name: "", token: "", newPassword: "" }) {
    const [values, setValues] = React.useState(inputValues);
  
    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const { value, name } = event.target;
      setValues({ ...values, [name]: value });
    };
  
    return { values, handleChange, setValues };
  }