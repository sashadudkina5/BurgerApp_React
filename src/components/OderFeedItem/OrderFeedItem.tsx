import styles from "./OderFeedItem.module.css";
import { CurrencyIcon } from "@ya.praktikum/react-developer-burger-ui-components";
import {
  getAllCreatedOrders,
  getListOfIngredients,
  isWSLoading,
  getWSError,
} from "../../redux_services/selectors";
import { useAppSelector, useAppDispatch } from "../../hooks/dispatch-selectos";
import moment from "moment";
import "moment/locale/ru";
import { TOrder } from "../../utils/types";
import React, { MouseEvent } from "react";
import { showDoneOrderDetails } from "../DoneOrderDetails/actions";
import { Link, useLocation } from "react-router-dom";

/**
 * Reusable component that renders each order in the list. Orders are fetched from WebSocket.
 * Users can click on an order to view detailed information in a modal.
 * Handles loading states and displays appropriate messages for loading and connection errors.
 *
 * @component
 */
function OrderFeedItem() {
  let location = useLocation();
  const dispatch = useAppDispatch();

  /**
   * Last 50 orders recieved from WebSocket. All statuses. Stored in redux
   */
  const allCreatedOrders = useAppSelector(getAllCreatedOrders);

  /**
   * Array of all ingredientes available. Stores in redux
   */
  const allIngredients = useAppSelector(getListOfIngredients);

  /**
   * True if WebSocket connection is loading. Stored in redux
   */
  const WSLoading = useAppSelector(isWSLoading);

  /**
   * Error in WebSocket response. Stored in redux
   */
  const connectionError = useAppSelector(getWSError);

  if (WSLoading) {
    return <p className="text text_type_main-default">Loading...</p>;
  }

  // Displays a message when there are connection errors
  if (connectionError !== "") {
    return (
      <p className="text text_type_main-default">
        Ошибка подключения. Повторите попытку позже
      </p>
    );
  }

  // Displays a message when there are no orders.
  if (!allCreatedOrders || allCreatedOrders.length === 0) {
    return <p className="text text_type_main-default">Заказов пока нет</p>;
  }

  /**
   * Function fetches clicked order into redux store.
   * The modal with order details is opened.
   *
   * @param event
   * @param selectedOrder - an order clicked by user
   */
  const handleClick = (
    event: MouseEvent<HTMLLIElement>,
    selectedOrder: TOrder
  ) => {
    dispatch(showDoneOrderDetails(selectedOrder));
  };

  /**
   * Array of [name of ingredient, its price] elements.
   */
  const ingredientPricesMap = allIngredients.ingredientsData.data
    .filter((ingredient) => ingredient.price !== undefined)
    .map(
      (ingredient) => [ingredient._id, ingredient.price] as [string, number]
    );

  /**
   * Array of order object and its total price
   */
  const ordersWithTotalPrice = allCreatedOrders.map((order) => {
    const totalPrice = order.ingredients.reduce((sum, ingredientId) => {
      const ingredientPrice = ingredientPricesMap.find(
        ([_id]) => _id === ingredientId
      )?.[1];

      if (ingredientPrice !== undefined) {
        return sum + ingredientPrice;
      } else {
        console.warn(`Price not found for ingredient with ID ${ingredientId}`);
        return sum;
      }
    }, 0);

    return {
      order,
      totalPrice,
    };
  });

  /**
   * Array of images of all ingredients available
   */
  const imageMap = allIngredients.ingredientsData.data.reduce(
    (map, ingredient) => {
      map[ingredient._id as string] = ingredient.image;
      return map;
    },
    {} as Record<string, string | undefined>
  );

  /**
   * Array of order object and its ingredients images
   */
  const ordersWithImages = allCreatedOrders.map((order) => {
    const images = order.ingredients.map(
      (ingredientId: string) => imageMap[ingredientId]
    );

    return {
      order,
      images,
    };
  });

  return (
    <>
      {ordersWithTotalPrice.map(({ order, totalPrice }) => (
        <Link
          to={`/feed/${order.number}`}
          state={{ backgroundLocation: location }}
          style={{
            textDecoration: "none",
            color: "inherit",
          }}
          key={order._id}
        >
          <li
            className={styles.feed_section_item}
            key={order.number}
            onClick={(event) => handleClick(event, order)}
          >
            <div className={styles.order_info_wrapper}>
              <p className="text text_type_main-default">{`#${order.number}`}</p>
              <p className="text text_type_main-default text_color_inactive">
                <time dateTime={moment(order.createdAt).format()}>
                  {moment(order.createdAt).calendar(null, {
                    sameDay: "[Сегодня], LT",
                    lastDay: "[Вчера], LT",
                    lastWeek: "DD.MM.YYYY, LT",
                    sameElse: "DD.MM.YYYY, LT",
                  })}
                </time>
              </p>
            </div>
            <h2 className="text text_type_main-medium">{order.name}</h2>
            <div className={styles.order_info_wrapper}>
              <div className={styles.images_container}>
                {ordersWithImages
                  .find((item) => item.order === order)
                  ?.images.map((imageUrl, index) => (
                    <div
                      className={styles.imageWrapper}
                      key={index}
                      style={{
                        zIndex: ordersWithImages.length + index,
                      }}
                    >
                      {imageUrl && (
                        <img
                          className={styles.ingredient_image}
                          src={imageUrl}
                          alt={`Ingredient ${index}`}
                        />
                      )}
                    </div>
                  ))}
              </div>
              <div className={styles.order_info_wrapper}>
                <p className="text text_type_digits-default">{totalPrice}</p>
                <CurrencyIcon type="primary" />
              </div>
            </div>
          </li>
        </Link>
      ))}
    </>
  );
}

export default OrderFeedItem;
