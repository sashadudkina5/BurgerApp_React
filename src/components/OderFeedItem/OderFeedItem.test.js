import React from 'react';
import { render } from '@testing-library/react';
import {OrderFeedItem} from './OrderFeedItem';
import { useAppSelector } from '../../hooks/dispatch-selectos';
import { useLocation } from 'react-router-dom';

import {
    getAllCreatedOrders,
    getListOfIngredients,
  } from "../../redux_services/selectors";

jest.mock('react-router-dom', () => ({
  useLocation: jest.fn(),
}));
jest.mock('../../hooks/dispatch-selectos', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: () => jest.fn()
}));

describe('OrderFeedItem', () => {
  it('renders correctly and matches snapshot', () => {

    const mockOrders = [
        {
          _id: '65b117ac87899c001b82b2a4',
          ingredients: [
            '643d69a5c3f7b9001cfa093d',
            '643d69a5c3f7b9001cfa0943',
            '643d69a5c3f7b9001cfa0942',
            '643d69a5c3f7b9001cfa0945',
            '643d69a5c3f7b9001cfa093d'
          ],
          status: 'done',
          name: 'Антарианский spicy флюоресцентный space бургер',
          createdAt: '2024-01-24T13:59:08.544Z',
          updatedAt: '2024-01-24T13:59:08.938Z',
          number: 32809
        },
        {
          _id: '65b1173d87899c001b82b2a0',
          ingredients: [
            '643d69a5c3f7b9001cfa093d',
            '643d69a5c3f7b9001cfa0943',
            '643d69a5c3f7b9001cfa0945',
            '643d69a5c3f7b9001cfa0943',
            '643d69a5c3f7b9001cfa093d'
          ],
          status: 'done',
          name: 'Антарианский флюоресцентный space бургер',
          createdAt: '2024-01-24T13:57:17.432Z',
          updatedAt: '2024-01-24T13:57:17.969Z',
          number: 32808
        }];
    const mockIngredients =  
        [
          {
            _id: '643d69a5c3f7b9001cfa093c',
            name: 'Краторная булка N-200i',
            type: 'bun',
            proteins: 80,
            fat: 24,
            carbohydrates: 53,
            calories: 420,
            price: 1255,
            image: 'https://code.s3.yandex.net/react/code/bun-02.png',
            image_mobile: 'https://code.s3.yandex.net/react/code/bun-02-mobile.png',
            image_large: 'https://code.s3.yandex.net/react/code/bun-02-large.png',
            __v: 0
          },
          {
            _id: '643d69a5c3f7b9001cfa0941',
            name: 'Биокотлета из марсианской Магнолии',
            type: 'main',
            proteins: 420,
            fat: 142,
            carbohydrates: 242,
            calories: 4242,
            price: 424,
            image: 'https://code.s3.yandex.net/react/code/meat-01.png',
            image_mobile: 'https://code.s3.yandex.net/react/code/meat-01-mobile.png',
            image_large: 'https://code.s3.yandex.net/react/code/meat-01-large.png',
            __v: 0
          },
          {
            _id: '643d69a5c3f7b9001cfa093e',
            name: 'Филе Люминесцентного тетраодонтимформа',
            type: 'main',
            proteins: 44,
            fat: 26,
            carbohydrates: 85,
            calories: 643,
            price: 988,
            image: 'https://code.s3.yandex.net/react/code/meat-03.png',
            image_mobile: 'https://code.s3.yandex.net/react/code/meat-03-mobile.png',
            image_large: 'https://code.s3.yandex.net/react/code/meat-03-large.png',
            __v: 0
          }];
    
    useAppSelector.mockImplementation((selector) => {
      if (selector === getAllCreatedOrders) return mockOrders;
      if (selector === getListOfIngredients) return { ingredientsData: { data: mockIngredients } };
      return undefined;
    });
    useLocation.mockReturnValue({pathname: '/feed/id123',});

    const { asFragment } = render(<OrderFeedItem />);

    expect(asFragment()).toMatchSnapshot();
  });
});
