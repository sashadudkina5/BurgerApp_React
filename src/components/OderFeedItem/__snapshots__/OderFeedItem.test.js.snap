// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderFeedItem renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <p
    class="text text_type_main-default"
  >
    Ошибка подключения. Повторите попытку позже
  </p>
</DocumentFragment>
`;
